### PR TITLE
Support for fuel gauge sensor MAX17048

### PR DESCRIPTION
**Pull request by maxpowel**
Hi, I am using this chip in my projects and I think that it is mature enought to be shared. And hopefully it can be useful for someone.

I also added a simple example. I tested it on NRF52 but it should work anywhere because it just requires a regular I2C.

### CODE REVIEW

**Comment by teburd** [comment]
I'd note that Zephyr now has a fuel gauge API that this would better be represented by I believe. I don't think this is stale so removing that regardless.

**Comment by teburd** [comment]
@aaronemassey please take a look and maybe make some suggestions on if this could be ported to the new fuel gauge api

**Comment by aaronemassey** [comment]
> @aaronemassey please take a look and maybe make some suggestions on if this could be ported to the new fuel gauge api

It looks like all of the exposed sensor channels (one of the sensor channels doesn't exist anymore) already exist as a fuel gauge property. So I don't see any reason why this couldn't be ported to the fuel gauge API with relative ease.

I'd intended to do much of the driver migration myself (should be straight forward refactor) with reviews from their active maintainers.

File: drivers/sensor/max17048/max17048.c [hunk 1]
```diff
@@ -0,0 +1,188 @@
+/* max17048.c - Driver for max17048 battery fuel gauge */
+
+/*
+ * Copyright (c) 2022 Alvaro Garcia Gomez <maxpowel@gmail.com>
```
Comment (aaronemassey) [inline drivers/sensor/max17048/max17048.c:4]: nit: 2023 copyright :stuck_out_tongue:

File: drivers/sensor/max17048/max17048.c [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+			val->val2 = decimal_part;
+		break;
+
+		case SENSOR_CHAN_GAUGE_VOLTAGE:
+			whole_part = data->voltage;
+			decimal_part = (data->voltage - whole_part) * 100 * 10000;
+
+			val->val1 = whole_part;
+			val->val2 = decimal_part;
+		break;
```
Comment (aaronemassey) [inline drivers/sensor/max17048/max17048.c:125]: The formatting for the break statements looks a little weird. Could you run a `clang-format`?

File: drivers/sensor/max17048/max17048.c [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+{
+	const struct max17048_config *cfg = dev->config;
+
+	if (!device_is_ready(cfg->i2c.bus)) {
+		LOG_ERR("Bus device is not ready");
+		return -ENODEV;
+	}
+
+	uint16_t version = max17048_read_register(dev, REGISTER_VERSION) & 0xFFF0;
+	if (version != 0x10){
```
Comment (aaronemassey) [inline drivers/sensor/max17048/max17048.c:156]: Seems like the version should be a device-tree property.

File: drivers/sensor/max17048/max17048.c [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+		printk("Wrong MAX17048 version, expected 0x10 but got %d. Maybe your wiring is wrong or it is a fake chip\n", version);
+		return -1;
```
Comment (aaronemassey) [inline drivers/sensor/max17048/max17048.c:158]: Returning -1 might imply `-EPERM` error which doesn't quite make sense here. How about `-ENOTSUP`?


See `errno.h`

File: drivers/sensor/max17048/max17048.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (c) 2022 Alvaro Garcia Gomez <maxpowel@gmail.com>
```
Comment (aaronemassey) [inline drivers/sensor/max17048/max17048.h:2]: nit: Copyright here too.

File: samples/sensor/max17048/Kconfig [hunk 6]
```diff
@@ -0,0 +1,16 @@
+# Copyright (c) 2022, Alvaro Garcia <maxpowel@gmail.com>
```
Comment (aaronemassey) [inline samples/sensor/max17048/Kconfig:1]: I don't think we need or want any of these KConfigs? None of the other drivers overwrite kconfigs.

**Review by aaronemassey — CHANGES_REQUESTED** [review]
Not a hard requirement, but I personally would prefer if you added emulated testing here. Based on the datasheet you can reuse a lot of the code from the `emul_sbs_gauge.c` since the `max17048` uses 16 byte words just like smbus.

File: include/zephyr/drivers/sensor.h [hunk 7]
```diff
@@ -168,6 +168,8 @@ enum sensor_channel {
 	SENSOR_CHAN_GAUGE_TIME_TO_EMPTY,
 	/** Time to full in minutes **/
 	SENSOR_CHAN_GAUGE_TIME_TO_FULL,
+	/** If the battery is being charged or discharged */
+	SENSOR_CHAN_GAUGE_CHARGING_STATE,
```
Comment (aaronemassey) [inline include/zephyr/drivers/sensor.h:172]: I don't think we want or need this. We can just check the charging current property.

**Review by aaronemassey — CHANGES_REQUESTED** [review]

**Comment by maxpowel** [comment]
Thank you for your time. I would be glad to adapt it to the new API taking into account all your comments. 

I was reading these links about the new api
https://docs.zephyrproject.org/latest/hardware/peripherals/fuel_gauge.html
https://github.com/zephyrproject-rtos/zephyr/blob/6cdef0894fe7aced7f58b1f34fa1ca2b691f56cb/include/zephyr/drivers/fuel_gauge.h
https://github.com/zephyrproject-rtos/zephyr/blob/6cdef0894fe7aced7f58b1f34fa1ca2b691f56cb/drivers/fuel_gauge/sbs_gauge/sbs_gauge.h

But I'm not sure about how to implement this api. Is there any example that I can use for reference?

Thank you

**Comment by aaronemassey** [comment]
> Thank you for your time. I would be glad to adapt it to the new API taking into account all your comments.
> 
> I was reading these links about the new api https://docs.zephyrproject.org/latest/hardware/peripherals/fuel_gauge.html https://github.com/zephyrproject-rtos/zephyr/blob/6cdef0894fe7aced7f58b1f34fa1ca2b691f56cb/include/zephyr/drivers/fuel_gauge.h https://github.com/zephyrproject-rtos/zephyr/blob/6cdef0894fe7aced7f58b1f34fa1ca2b691f56cb/drivers/fuel_gauge/sbs_gauge/sbs_gauge.h
> 
> But I'm not sure about how to implement this api. Is there any example that I can use for reference?
> 
> Thank you

Working prototype for the [SBS Compliant Fual Gauge driver](https://github.com/zephyrproject-rtos/zephyr/blob/513e03ad684810832968cfc8ed249b1898c6e92c/drivers/fuel_gauge/sbs_gauge/sbs_gauge.c)

Reminder: This API is still in the experimental stage, so some things are likely to change. I don't mind being responsible for those changes and migrations though.

File: samples/sensor/max17048/Kconfig (hunk 6 above)
Comment (maxpowel) [reply samples/sensor/max17048/Kconfig:1]: I agree

**Comment by maxpowel** [comment]
> Reminder: This API is still in the experimental stage, so some things are likely to change. I don't mind being responsible for those changes and migrations though.
No worries, I know that is is experimental and it is no problem if I have to do more changes in the future.

I have just adapted the code to the fuel gague api, updated the sample and added emul tests. I still want to improve a bit the code and the tests but can you tell me if the api is correctly implemented (in terms of structure)? 

I would be grateful if you can also take a look on the sample and tell me if this is how the fuel gauge api should be used.
https://github.com/zephyrproject-rtos/zephyr/blob/16043f62a40ab6c5f807d9dec5bc2f866bcf08c2/samples/fuel_gauge/max17048/src/main.c
Basically I call the function fuel_gauge_get_prop with a list of propts and then read the result.

Thanks

File: drivers/fuel_gauge/max17048/emul_max17048.c [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+			default:
+				LOG_ERR("Unexpected msg1 length %d", msgs->len);
+				return -EIO;
+			}
+		} else {
+			/* We write a word (2 bytes by the SBS spec) */
+			if (msgs->len != 2) {
+				LOG_ERR("Unexpected msg1 length %d", msgs->len);
+			}
+			uint16_t *value = (uint16_t *)msgs->buf;
```
Comment (aaronemassey) [inline drivers/fuel_gauge/max17048/emul_max17048.c:128]: Change

```
uint16_t *value = (uint16_t *)msgs->buf;
```
to
```
uint16_t value = sys_get_le16(msgs->buf);
```

This was a bug in the original emul_sbs_gauge emulator that I only found after using a big endian qemu.

See https://github.com/zephyrproject-rtos/zephyr/pull/55103

Thank you for contributing an emulator!!!

**Review by aaronemassey — CHANGES_REQUESTED** [review]
This is great! Really appreciate the contribution of an additional fuel gauge driver and associated emulator.

**Comment by aaronemassey** [comment]
> I would be grateful if you can also take a look on the sample and tell me if this is how the fuel gauge api should be used. https://github.com/zephyrproject-rtos/zephyr/blob/16043f62a40ab6c5f807d9dec5bc2f866bcf08c2/samples/fuel_gauge/max17048/src/main.c Basically I call the function fuel_gauge_get_prop with a list of propts and then read the result.
> 
> Thanks

Left some fairly minor comments. This is great, thank you!

File: tests/drivers/fuel_gauge/max17048/prj.conf [hunk 9]
```diff
@@ -0,0 +1,8 @@
+CONFIG_ZTEST=y
+CONFIG_ZTEST_NEW_API=y
+CONFIG_I2C=y
+CONFIG_TEST_USERSPACE=y
+CONFIG_LOG=y
+CONFIG_SENSOR=y
```
[the remark this answers is not in the document]
Comment (maxpowel) [reply tests/drivers/fuel_gauge/max17048/prj.conf:6]: I copied it from `https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/fuel_gauge/sbs_gauge/prj.conf` which has `CONFIG_SENSOR`.
It compiles perfectly without this so maybe in is not necessary there either.

Anyway, I removed it from this and after a few changes and everthing is ok

**Comment by maxpowel** [comment]
I finished all the improvements I wanted to do. I also tested the example with a real device and it works perfectly!

File: drivers/fuel_gauge/max17048/max17048.c [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+
+/* % per hour, positive is charging or negative if discharging */
+int max17048_crate(const struct device *i2c_dev, float *response) {
+	int16_t value;
+	int rc = max17048_read_register(i2c_dev, REGISTER_CRATE, &value);
+	if (rc < 0) {
+		return rc;
+	}
+
+	*response = (float)value * 0.208f;
```
Comment (aaronemassey) [inline drivers/fuel_gauge/max17048/max17048.c:97]: Can we also document `0.208f`?

File: samples/fuel_gauge/max17048/README.rst [hunk 11]
```diff
@@ -0,0 +1,51 @@
+.. _MAX17048_sample:
+
+MAX17048 Li-Ion battery fuel gauge
+###################################
+
+Overview
+********
+
+This sample shows how to use the Zephyr :ref:`sensor_api` API driver for the `MAX17048` sensor.
```
Comment (aaronemassey) [inline samples/fuel_gauge/max17048/README.rst:9]: Should update this to `:ref:fuel_gauge_api`

File: drivers/fuel_gauge/max17048/emul_max17048.c [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+	uint16_t mfr_acc;
+};
+
+/** Static configuration for the emulator */
+struct max17048_emul_cfg {
+	/** I2C address of emulator */
+	uint16_t addr;
+};
+
+static int emul_max17048_reg_write(const struct emul *target, int reg, int val)
```
Comment (aaronemassey) [inline drivers/fuel_gauge/max17048/emul_max17048.c:35]: Let's remove the commented out code and probably just return -EIO if we're not expecting a write

File: drivers/fuel_gauge/max17048/emul_max17048.c [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+#include <zephyr/drivers/emul.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/i2c_emul.h>
+#include <zephyr/sys/byteorder.h>
+
+#include "max17048.h"
+
+/** Run-time data used by the emulator */
+struct max17048_emul_data {
+	uint16_t mfr_acc;
```
Comment (aaronemassey) [inline drivers/fuel_gauge/max17048/emul_max17048.c:26]: We can delete the max17048_emul_data since it's unused.

File: drivers/fuel_gauge/max17048/max17048.c [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+struct max17048_data {
+	float charge;
+	float voltage;
+	uint16_t time_to_full;
+	uint16_t time_to_empty;
+	bool charging;
+};
+
+/* Read buffer */
+unsigned char max17048_buffer[2];
```
Comment (aaronemassey) [inline drivers/fuel_gauge/max17048/max17048.c:39]: Since we only use this in the `max17048_read_register()` function, let's make it local to that function.

File: samples/fuel_gauge/max17048/README.rst [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+
+Building and Running
+********************
+
+The sample can be configured to support MAX17048 sensors connected via either I2C. It only needs
+an I2C pin configuration
+
+Features
+********
+By using this sensor you can get the following information:
```
Comment (aaronemassey) [inline samples/fuel_gauge/max17048/README.rst:23]: Fuel guage is technically a sensor but might be a good idea to just change `sensor` to `fuel gauge` here in case someone confuses it with the `sensor` API.

File: samples/fuel_gauge/max17048/README.rst [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+if you plug/unplug a charger it will take some time until it is actually detected by the chip. Don't try to plug/unplug
+to see in real time the charging status change because it will not work. If you really need to know exactly the moment
+when the battery is being charged you will need other method.
+
+Sample output
+*************
+
+```
+*** Booting Zephyr OS build 16043f62a40a ***
+Found device "max17048@36", getting sensor data
```
Comment (aaronemassey) [inline samples/fuel_gauge/max17048/README.rst:42]: `getting sensor` to `getting fuel gauge`, again just to avoid confusion with the sensor api.

**Review by aaronemassey — CHANGES_REQUESTED** [review]
Really close! Just a few minor things. Also feel free to ping `aaronmassey#4455` on the Zephyr Discord if you have any questions and want a usually quicker response.

One thing I noticed was the usage of `floats` and many MCUs can't work with floats. But IMO that's not a blocker for the PR because it can always be changed in the future since the API works with ints.

File: drivers/fuel_gauge/max17048/max17048.c [hunk 17]
```diff
@@ -44,37 +43,59 @@ unsigned char max17048_buffer[2];
...
 /* Battery voltage */
-float max17048_voltage(const struct device *i2c_dev) {
-	return (float) max17048_adc(i2c_dev)  * 78.125 / 1000000;
+int max17048_voltage(const struct device *i2c_dev, float *response) {
+	uint16_t data;
+	int rc = max17048_adc(i2c_dev, &data);
+	if (rc < 0){
+		return rc;
+	}
+	*response = (float)data * 78.125 / 1000000;
```
[the remark this answers is not in the document]
Comment (maxpowel) [reply drivers/fuel_gauge/max17048/max17048.c:71]: This thing comes from de datasheet
https://www.analog.com/media/en/technical-documentation/data-sheets/MAX17048-MAX17049.pdf
Page 10, Table 2. Register Summary:
78.125µV/cell
This ugly float operation is required when converting the value from the register into a voltage integer in the api shape.

Using integers would reduce precision

I will add a reference there as a comment for sure. I will review the code for more things like this

File: drivers/fuel_gauge/max17048/max17048.c [hunk 18]
```diff
@@ -44,37 +43,59 @@ unsigned char max17048_buffer[2];
...
-	return register_value / 256;
+int max17048_percent(const struct device *i2c_dev, float *response) {
+	uint16_t data;
+	int rc = max17048_read_register(i2c_dev, REGISTER_SOC, &data);
+	if (rc < 0) {
+		return rc;
+	}
+	*response = (float) data;
+
+	*response /= 256;
```
[the remark this answers is not in the document]
Comment (maxpowel) [reply drivers/fuel_gauge/max17048/max17048.c:85]: This thing comes from de datasheet in the same place as above
https://www.analog.com/media/en/technical-documentation/data-sheets/MAX17048-MAX17049.pdf
Page 10, Table 2. Register Summary:
1%/256 

I totally agree with you, people should not need to check the datasheet to understand how does it work. I will add it as a comment too

File: drivers/fuel_gauge/max17048/max17048.c (hunk 10 above)
Comment (maxpowel) [reply drivers/fuel_gauge/max17048/max17048.c:97]: Again, this weird values just comes from the datasheet
https://www.analog.com/media/en/technical-documentation/data-sheets/MAX17048-MAX17049.pdf
Page 11
Table 2. Register Summary (continued)
0.208%/hr

Sure I will document it.

**Comment by stephanosio** [comment]
> In any case, I did not want to reinvent the wheel implementing a software float processing because there are probably other easier options since this problem should be common

For most architectures, software floating point support is already implemented by the toolchain and there is no need for you to manually provide one.

**Comment by maxpowel** [comment]
Thanks for the info @stephanosio

Shame on me, but just doing bit shifting I can avoid all the floats. The values sent by the IC are just integer values and only the constants are float

**Comment by maxpowel** [comment]
I applied all the changes. Basically removed all float operations, cleaned coded based on your suggestions, added more documentation and updated the sensor references in the doc.

Everything compiles and run as expected (also tested in my devices)

File: drivers/fuel_gauge/max17048/max17048.c [hunk 19]
```diff
@@ -62,39 +67,57 @@ int max17048_adc(const struct device *i2c_dev, uint16_t *response) {
...
 		return rc;
 	}
 
-	*response = (float)value * 0.208f;
+	/* Once the value is read, it has to ve converted to something useful. The datasheet
+	https://www.analog.com/media/en/technical-documentation/data-sheets/MAX17048-MAX17049.pdf
+	Page 11, Table 2. Register Summary (continued): 0.208%/hr
+	
+	To avoid floats, the value will be multiplied by 208 instead of 0.208, taking into account
+	that the value will ve 1000 times higher
```
Comment (aaronemassey) [inline drivers/fuel_gauge/max17048/max17048.c:129]: nit: `ve` to `be` :stuck_out_tongue:

File: drivers/fuel_gauge/max17048/max17048.c [hunk 20]
```diff
@@ -39,17 +39,14 @@ struct max17048_data {
...
-
 /**
  * I2C communication
  * The way we read a value is first writing the address we want to read and then
  * wait for 2 bytes containing the data.
  */
 int max17048_read_register(const struct device *dev, uint8_t registerId, uint16_t *response)
 {
-
+	unsigned char max17048_buffer[2];
```
Comment (aaronemassey) [inline drivers/fuel_gauge/max17048/max17048.c:51]: Bah, I should've caught this earlier sorry. Can we make this a `uint8_t` array? That seems to be the style I've seen in Zephyr.

**Review by aaronemassey — CHANGES_REQUESTED** [review]
Awesome! One minor item that I missed last time.

After a bit of rebasing to fix commit history, I think we can get this merged pretty soon assuming @teburd approves.

**Comment by aaronemassey** [comment]
Thanks for your patience with my reviews, I'm still getting used to GitHub's review format from Gerrit.

File: drivers/fuel_gauge/max17048/max17048.c (hunk 20 above)
Comment (aaronemassey) [reply drivers/fuel_gauge/max17048/max17048.c:51]: It's effectively the same, so I don't think there's any reason to retest.

File: drivers/fuel_gauge/max17048/max17048.c (hunk 19 above)
Comment (maxpowel) [reply drivers/fuel_gauge/max17048/max17048.c]: Oh, I checked it twice. Looks that I'm blind

**Comment by maxpowel** [comment]
> Thanks for your patience with my reviews, I'm still getting used to GitHub's review format from Gerrit.

No worries, take your time. There is not reason to rush. Thank you for your reviews.

I have just fixed your last rewiews.

**Review by aaronemassey — APPROVED** [review]
Lgtm, going to need to rebase the fixup commits and get @teburd 's view.

Thank you for the contribution and work!

I'll probably have to lgtm again after you rebase too.

**Comment by maxpowel** [comment]
Thank you for your time, please tell me if you need something else from me.

By the way, just as a reminder, FUEL_GAUGE_INIT will need to be added once it is ready.

**Comment by aaronemassey** [comment]
> By the way, just as a reminder, FUEL_GAUGE_INIT will need to be added once it is ready.

https://github.com/zephyrproject-rtos/zephyr/pull/55887 will probably merge fairly quick.

> please tell me if you need something else from me.

Main things I see blocking a merge are:

* One more +1 from another contributor, probably teburd.
* Rebase commits in branch to remove squash fixup commits for a cleaner history
* Rebase commits to remove all merge commits (Zephyr does rebasing over merging for a linear commit history)

**Review by aaronemassey — CHANGES_REQUESTED** [review]
See https://github.com/zephyrproject-rtos/zephyr/pull/52349#issuecomment-1472319417

**Comment by maxpowel** [comment]
I did the rebase stuff and also added the CONFIG_FUEL_GAUGE_INIT
Now I am dealing with all the style faults to pass all the workflows

**Comment by maxpowel** [comment]
Ok now my code passes the compliance checks although there is an error related to "ZEPHYR_LIBLC3_MODULE" but I thinkg that this not related to me.

**Review by aaronemassey — APPROVED** [review]

**Review by teburd — APPROVED** [review]
Looks good to me

If there's a sample per fuel gauge it'll get unwieldy pretty quickly as sensors saw, but I don't have a great recommendation other than perhaps as more come in create a common sample that uses overlays or maybe the future snippets feature @mbolivar is working on I think.

Board overlays seem like they might be hard here as most boards don't have a fuel gauge, its something you want to test out with like a breakout board probably.
